lib/rouge/lexers/coq.rb

```diff
@@ -82,16 +76,105 @@ def self.classify(x)
+      id_first = /\p{L}/
+      id_first_underscore = /(?:\p{L}|_)/
+      id_subsequent = /(?:\p{L}|\p{N}|_|')/ # a few missing? some mathematical ' primes and subscripts
+      id = /(?:#{id_first}#{id_subsequent}*)|(?:#{id_first_underscore}#{id_subsequent}+)/i
+      dot_id = /\.(#{id})/i
       dot_space = /\.(\s+)/
+      proof = /Proof(\s*).(\s+)/i
+      qed = /(Qed|Defined|Save|Admitted)(\s*).(\s+)/i
       module_type = /Module(\s+)Type(\s+)/
       set_options = /(Set|Unset)(\s+)(Universe|Printing|Implicit|Strict)(\s+)(Polymorphism|All|Notations|Arguments|Universes|Implicit)(\s*)\./m
 
       state :root do
-        rule %r/[(][*](?![)])/, Comment, :comment
+        mixin :begin_proof
+        mixin :sentence
+      end
+
+      state :sentence do
+        mixin :comment_whitespace
+        mixin :module_setopts
+        # After parsing the id, end up in sentence_postid
+        rule id do |m|
+          @name = m[0]
+          @id_dotted = false
+          push :sentence_postid
+          push :continue_id
+        end
+      end
+
+      state :begin_proof do
+        rule proof do |m|
+          token Keyword, "Proof"
+          token Text::Whitespace, m[1]
+          token Punctuation::Indicator, '.'
+          token Text::Whitespace, m[2]
+          push :proof_mode
+        end
+      end
+
+      state :proof_mode do
+        mixin :comment_whitespace
+        mixin :module_setopts
+        mixin :begin_proof
+        rule qed do |m|
+          token Keyword, m[1]
+          token Text::Whitespace, m[2]
+          token Punctuation::Indicator, '.'
+          token Text::Whitespace, m[3]
+          pop! # :proof_mode
+        end
+        # the whole point of parsing Proof/Qed, normally some of these will be operators
+        rule %r/(?:\-+|\++|\*+)/, Punctuation
+        rule %r/[{}]/, Punctuation
+        # toplevel_selector
+        rule %r/(!|all|par):/ do |m|
+          token Keyword::Pseudo, m[1]
+          token Punctuation, ':'.dup
+        end
+        # numbered goals 1: {} 1,2: {}
+        rule %r/\d+/, Num::Integer, :numeric_labels
+        # [named_goal]: { ... }
+        rule %r/\[(\s*)(#{id})(\s*)(\])(\s*)(:)/ do |m|
+          token Punctuation, '['
+          token Text::Whitespace, m[1]
+          token Name::Constant, m[2]
+          token Text::Whitespace, m[3]
+          token Punctuation, ']'.dup # without dup, you get an error when you attempt to appending to it with ':'
+          token Text::Whitespace, m[5]
+          token Punctuation, ':'.dup
+        end
+        # After parsing the id, end up in sentence_postid
+        rule id do |m|
+          @name = m[0]
+          @id_dotted = false
+          push :sentence_postid
+          push :continue_id
+        end
+      end
+
+      state :numeric_labels do
+        mixin :whitespace
+        rule %r(,(\s*)(\d+)) do |m|
+          token Punctuation, ','
+          token Text::Whitespace, m[1]
+          token Num::Integer, m[2]
+        end
```

Comment:
**suggestion:** Similar to above, we can include punctuation in the regex and use `groups` to simplify the rule. WDYT?

```suggestion
        rule %r/(,)(\s*)(\d+)/ do |m|
          groups Punctuation, Text::Whitespace, Num::Integer
        end
```